mullvad-relay-selector/src/relay_selector/matcher.rs

```diff
@@ -0,0 +1,288 @@
+        bridge_state: BridgeState,
+        wireguard_data: &'a WireguardEndpointData,
+        custom_lists: &CustomListsSettings,
+    ) -> RelayMatcher<AnyTunnelMatcher<'a>> {
+        let endpoint_matcher = AnyTunnelMatcher {
+            wireguard: WireguardMatcher::new(query.wireguard_constraints.clone(), wireguard_data),
+            openvpn: OpenVpnMatcher::new(
+                query.openvpn_constraints.clone(),
+                openvpn_data,
+                bridge_state,
+            ),
+            tunnel_type: query.tunnel_protocol,
+        };
+        Self::using(query, custom_lists, endpoint_matcher)
+    }
+}
+
+impl<T: EndpointMatcher> RelayMatcher<T> {
+    pub fn using(
+        query: RelayQuery,
+        custom_lists: &CustomListsSettings,
+        endpoint_matcher: T,
+    ) -> RelayMatcher<T> {
+        RelayMatcher {
+            locations: ResolvedLocationConstraint::from_constraint(query.location, custom_lists),
+            providers: query.providers,
+            ownership: query.ownership,
+            endpoint_matcher,
+        }
+    }
+
+    /// Filter a list of relays and their endpoints based on constraints.
+    /// Only relays with (and including) matching endpoints are returned.
+    pub fn filter_matching_relay_list<'a, R: Iterator<Item = &'a Relay> + Clone>(
+        &self,
+        relays: R,
+    ) -> Vec<Relay> {
+        let shortlist = relays
+            // Filter on active relays
+            .filter(|relay| filter_on_active(relay))
+            // Filter by location
+            .filter(|relay| filter_on_location(&self.locations, relay))
+            // Filter by ownership
+            .filter(|relay| filter_on_ownership(&self.ownership, relay))
+            // Filter by providers
+            .filter(|relay| filter_on_providers(&self.providers, relay))
+            // Filter on relay type & relay specific properties
+            .filter(|relay| self.endpoint_matcher.is_matching_relay(relay));
+
+        // The last filtering to be done is on the `include_in_country` attribute found on each
+        // relay. When the location constraint is based on country, a relay which has `include_in_country`
+        // set to true should always be prioritized over relays which has this flag set to false.
+        // We should only consider relays with `include_in_country` set to false if there are no
+        // other candidates left.
+        if !shortlist.clone().any(|relay| relay.include_in_country) {
+            shortlist.cloned().collect()
+        } else {
+            shortlist
+                .filter(|relay| filter_on_include_in_country(&self.locations, relay))
+                .cloned()
+                .collect()
+        }
+    }
+}
+
+/// EndpointMatcher allows to abstract over different tunnel-specific or bridge constraints.
+/// This enables one to not have false dependencies on OpenVpn specific constraints when
+/// selecting only WireGuard tunnels.
+pub trait EndpointMatcher {
```

Comment:
Are we actually using this trait anymore? It seems like we only create `RelayMatcher`s with `AnyTunnelMatcher`, or am I missing something?